test/extensions/filters/http/ext_authz/ext_authz_http_fuzz_test.cc

```diff
@@ -0,0 +1,143 @@
+                                  -> Http::AsyncClient::Request* {
+          if (status_ == "200") {
+            const auto headers = TestCommon::makeHeaderValueOption({{":status", status_, false}});
+            http_client_->onSuccess(http_sync_request_, TestCommon::makeMessageResponse(headers));
+          } else {
+            http_client_->onFailure(http_sync_request_, Http::AsyncClient::FailureReason::Reset);
+          }
+          return &http_sync_request_;
+        }));
+  }
+
+  std::unique_ptr<Filters::Common::ExtAuthz::RawHttpClientImpl>
+  newRawHttpClientImpl(const ExtAuthzTestCase::AuthResult result) {
+    http_client_ = new Filters::Common::ExtAuthz::RawHttpClientImpl(cm_, createConfig());
+    status_ = resultCaseToHttpStatus(result);
+    return std::unique_ptr<Filters::Common::ExtAuthz::RawHttpClientImpl>(http_client_);
+  }
+
+private:
+  Filters::Common::ExtAuthz::ClientConfigSharedPtr createConfig() {
+    envoy::extensions::filters::http::ext_authz::v3::ExtAuthz proto_config{};
+    const std::string default_yaml = R"EOF(
+      http_service:
+        server_uri:
+          uri: "ext_authz:9000"
+          cluster: "ext_authz"
+          timeout: 0.25s
+        authorization_request:
+          headers_to_add:
+          - key: "x-authz-header1"
+            value: "value"
+          - key: "x-authz-header2"
+            value: "value"
+        authorization_response:
+          allowed_upstream_headers:
+            patterns:
+            - exact: Bar
+              ignore_case: true
+            - prefix: "X-"
+              ignore_case: true
+          allowed_upstream_headers_to_append:
+            patterns:
+            - exact: Alice
+              ignore_case: true
+            - prefix: "Append-"
+              ignore_case: true
+          allowed_client_headers:
+            patterns:
+            - exact: Foo
+              ignore_case: true
+            - prefix: "X-"
+              ignore_case: true
+          allowed_client_headers_on_success:
+            patterns:
+            - prefix: "X-Downstream-"
+              ignore_case: true
+      )EOF";
+    TestUtility::loadFromYaml(default_yaml, proto_config);
+
+    return std::make_shared<Filters::Common::ExtAuthz::ClientConfig>(proto_config, 250, "/bar",
+                                                                     factory_context_);
+  }
+
+  NiceMock<Upstream::MockClusterManager> cm_;
+  NiceMock<Http::MockAsyncClient> internal_http_mock_client_;
+  NiceMock<Http::MockAsyncClientRequest> http_sync_request_;
+  NiceMock<Server::Configuration::MockServerFactoryContext> factory_context_;
+
+  // Set by calling newRawHttpClientImpl
+  std::string status_;
+  Filters::Common::ExtAuthz::RawHttpClientImpl* http_client_;
+};
+
+DEFINE_PROTO_FUZZER(const ExtAuthzTestCase& input) {
+  static ReusableFuzzerUtil fuzzer_util;
+  static ReusableHttpClientFactory http_client_factory;
```

Comment:
I am wondering what's the benefit to have a factory to create the filter instead of directly call a library function to create the filter, like you can put the code from https://github.com/envoyproxy/envoy/blob/08e23c4eb31ad0a642f415c8872d542364083b34/test/extensions/filters/http/ext_authz/ext_authz_fuzz_test.cc#L93 to https://github.com/envoyproxy/envoy/blob/08e23c4eb31ad0a642f415c8872d542364083b34/test/extensions/filters/http/ext_authz/ext_authz_fuzz_test.cc#L104 into the library function and call it.

Reply:
The factory makes it easier to run individual fuzzer test cases without recreating all the mocks each time. The mock call expectations just get set once and the data they return gets updated with each test call. Notice the fuzzer util is static in the actual fuzzer tests so it get reused.

Another benefit is that all the code that is shared between the grpc and http fuzzers is in this file instead of copied between the two new fuzzers.